consensus/hotstuff/pacemaker/pacemaker.go

```diff
@@ -12,111 +12,166 @@ import (
+	if err != nil {
+		return fmt.Errorf("could not persist liveness data: %w", err)
+	}
+
 	timerInfo := p.timeoutControl.StartTimeout(model.ReplicaTimeout, newView)
 	p.notifier.OnStartingTimeout(timerInfo)
-	return &model.NewViewEvent{View: p.currentView}
+	return nil
 }
 
 // CurView returns the current view
-func (p *NitroPaceMaker) CurView() uint64 {
-	return p.currentView
+func (p *ActivePaceMaker) CurView() uint64 {
+	return p.livenessData.CurrentView
 }
 
 // TimeoutChannel returns the timeout channel for current active timeout.
 // Note the returned timeout channel returns only one timeout, which is the current
 // timeout.
 // To get the timeout for the next timeout, you need to call TimeoutChannel() again.
-func (p *NitroPaceMaker) TimeoutChannel() <-chan time.Time {
+func (p *ActivePaceMaker) TimeoutChannel() <-chan time.Time {
 	return p.timeoutControl.Channel()
 }
 
 // ProcessQC notifies the pacemaker with a new QC, which might allow pacemaker to
-// fast forward its view.
-func (p *NitroPaceMaker) ProcessQC(qc *flow.QuorumCertificate) (*model.NewViewEvent, bool) {
-	if qc.View < p.currentView {
-		return nil, false
+// fast-forward its view.
+// No errors are expected, any error should be treated as exception
+func (p *ActivePaceMaker) ProcessQC(qc *flow.QuorumCertificate) (*model.NewViewEvent, error) {
+	if qc.View < p.CurView() {
+		return nil, nil
 	}
+
+	p.timeoutControl.OnProgressBeforeTimeout()
+
 	// qc.view = p.currentView + k for k ≥ 0
 	// 2/3 of replicas have already voted for round p.currentView + k, hence proceeded past currentView
 	// => 2/3 of replicas are at least in view qc.view + 1.
 	// => replica can skip ahead to view qc.view + 1
-	p.timeoutControl.OnProgressBeforeTimeout()
-
 	newView := qc.View + 1
+	err := p.updateLivenessData(newView, qc, nil)
+	if err != nil {
+		return nil, err
+	}
+
 	p.notifier.OnQcTriggeredViewChange(qc, newView)
-	return p.gotoView(newView), true
+	return &model.NewViewEvent{View: newView}, nil
 }
 
-func (p *NitroPaceMaker) ProcessTC(tc *flow.TimeoutCertificate) (*model.NewViewEvent, bool) {
-	panic("not yet implemented")
+// ProcessTC notifies the Pacemaker of a new timeout certificate, which may allow
+// Pacemaker to fast-forward its current view.
+// A nil TC is an expected valid input, so that callers may pass in e.g. `Proposal.LastViewTC`,
+// which may or may not have a value.
+// No errors are expected, any error should be treated as exception
+func (p *ActivePaceMaker) ProcessTC(tc *flow.TimeoutCertificate) (*model.NewViewEvent, error) {
+	if tc == nil || tc.View < p.CurView() {
+		return nil, nil
+	}
+
+	newView := tc.View + 1
+	err := p.updateLivenessData(newView, tc.NewestQC, tc)
+	if err != nil {
+		return nil, err
+	}
+
+	p.notifier.OnTcTriggeredViewChange(tc, newView)
+	return &model.NewViewEvent{View: newView}, nil
 }
 
-func (p *NitroPaceMaker) OnPartialTC(curView uint64) {
-	panic("not yet implemented")
+func (p *ActivePaceMaker) OnPartialTC(newView uint64) {
```

Comment:
I am starting to believe that the PaceMaker does not need to know about the Bracha timeout. Reasoning:
* We are building an information-driven system (specifically, because it helps modularizing algorithms). Information is idempotent. In this particular case, it means that reaching a timeout condition repeatedly should leave the system state invariant. 

   Benefits:
   * The pace maker can be only concerned with its time-based timeouts. No need to cancel the timeout if a Bracha timeout occurs first. Repeated timeout triggers for the same view don't hurt. 
   * If a Bracha timeout occurred, the Event Hander can directly call into SafetyRules. SafetyRules gracefully handles repeated calls for `ProduceTimeout`. 
      
        No need to call into PaceMaker to trigger the timeout, which then loops again through the Event Loop to trigger `SafetyRules`

Reply:
Yeah, my initial assumption was built on idea of canceling the timeout, but yeah, I will rollback changes for canceling timeout and just leave it as it is.